### ISSUE TITLE
Lint suggestion: write to raw pointer with drop glue

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust-clippy#4294]
One of the subtle aspects of `*ptr = val` where `ptr` is a raw pointer is that this will drop the old contents stored behind that pointer. On the one hand that might not be surprising because this is the same for mutable references, but on the other hand raw pointers are often used to point to invalid/uninitialized data, and then this is an easy way to cause a double-drop or to drop some garbage data. I have made this mistake myself and seen it a few times "out there".

So I think it'd be a good idea to lint against `*ptr = val` when `ptr: *mut T` and `T` has drop glue.  The lint should recommend to use `ptr.write(val)` instead, which does not drop. Code that *wants* to drop should explicitly do `ptr.drop_in_place()`.

**Comment by gnzlbg** [comment]
As in, the lint should suggest to re-write:

```rust
*ptr = val;
```

as 

```rust
ptr.drop_in_place();
ptr.write(val);
```

and contain a `note` that say something along these lines: "If the memory pointed to by `ptr` is not valid, the behavior of trying to drop that memory using `ptr.drop_in_place();` is undefined. This can happen if, for example, `*ptr = val` is used to initialize uninitialized memory".

**Comment by flip1995** [comment]
I would keep the note on the lint message short ("`ptr.drop_in_place()` could be undefined behavior") and put the long explanation in the documentation.

**Comment by Robert4S** [comment]
spent a bunch of time debugging *ptr = val where ptr is uninitialised, this would've helped a ton.

**Comment by y21** [comment]
During the FCP for the implementation of this, @kpreid made an [interesting point here](https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/FCP.3A.20.60raw_assign_to_drop.60/near/499777082) that the [proposed suggestion](https://github.com/rust-lang/rust-clippy/issues/4294#issuecomment-514160923) in this comment thread has subtly different semantics wrt unwinding: `*x = y;` will still write the new value even if dropping `*x` panics, which may or may not matter if you try to read from it before writing another value and could lead to UB. So you'd need to wrap the `drop_in_place` in a `catch_unwind` to really get the same semantics.

I was initially of the same opinion that doing these operations separately is better for the sake of making the drop effect obvious, but this convinced me that this is probably not a good suggestion.
